linera-witty/src/runtime/wasmtime/parameters.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Representation of Wasmtime function parameter types.
+
+use crate::{primitive_types::FlatType, Layout};
+use frunk::{hlist, hlist_pat, HCons, HList};
+use wasmtime::{WasmParams, WasmTy};
+
+/// Conversions between flat layouts and Wasmtime parameter types.
+pub trait WasmtimeParameters {
+    /// The type Wasmtime uses to represent the parameters.
+    type Parameters: WasmParams;
+
+    /// Converts from this flat layout into Wasmtime's representation.
+    fn into_wasmtime(self) -> Self::Parameters;
+
+    /// Converts from Wasmtime's representation into a flat layout.
+    fn from_wasmtime(parameters: Self::Parameters) -> Self;
+}
+
+/// Helper macro to implement [`WasmtimeParameters`] for flat layouts up to the maximum limit.
+///
+/// The maximum number of parameters is defined by the [canonical
+/// ABI](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#flattening)
+/// as the `MAX_FLAT_PARAMS` constant. There is no equivalent constant defined in Witty. Instead,
+/// any attempt to use more than the limit should lead to a compiler error. Therefore, this macro
+/// only implements the trait up to the limit. The same is done in other parts of the code, like
+/// for example in
+/// [`FlatHostParameters`][`crate::imported_function_interface::FlatHostParameters`].
+macro_rules! parameters {
+    ($( $names:ident : $types:ident ),*) => {
+        parameters!(| $( $names: $types ),*);
+    };
+
+    ($( $names:ident : $types:ident ),* |) => {
+        parameters!(@generate $( $names: $types ),*);
+    };
+
+    (
+        $( $names_to_generate:ident : $types_to_generate:ident ),* |
+        $next_name:ident : $next_type:ident $( , $queued_names:ident : $queued_types:ident )*
+    ) => {
+        parameters!(@generate $( $names_to_generate: $types_to_generate ),*);
+        parameters!(
+            $( $names_to_generate: $types_to_generate, )*
+            $next_name: $next_type | $( $queued_names: $queued_types ),*);
+    };
+
+    (@generate $( $names:ident : $types:ident ),*) => {
```

Comment:
nit: same comment on `@generate` as previous PRs

Reply:
Done, thanks. Sorry, I thought I had removed the last one with Wasmer :(